compiler/kernel/op.go

```diff
@@ -50,14 +50,15 @@ import (
 var ErrJoinParents = errors.New("join requires two upstream parallel query paths")
 
 type Builder struct {
-	rctx     *runtime.Context
-	mctx     *zed.Context
-	source   *data.Source
-	readers  []zio.Reader
-	progress *zbuf.Progress
-	arena    *zed.Arena // For zed.Values created during compilation.
-	deletes  *sync.Map
-	funcs    map[string]expr.Function
+	rctx      *runtime.Context
+	mctx      *zed.Context
+	source    *data.Source
+	readers   []zio.Reader
+	progress  *zbuf.Progress
+	arena     *zed.Arena // For zed.Values created during compilation.
+	deletes   *sync.Map
+	funcs     map[string]expr.Function
+	resetters expr.Resetters
```

Comment:
I think we might need an expr.Resetters stack here to handle stateful expressions nested at different levels correctly but this looks good for now.

Reply:
I think the one area for concern here is with over expressions. Everywhere else we just collect resetters for the expressions before recursing down into the flowgraph.